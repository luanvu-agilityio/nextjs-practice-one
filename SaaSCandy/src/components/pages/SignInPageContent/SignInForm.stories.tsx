import type { Meta, StoryObj } from '@storybook/nextjs';
import { SignInPageContent } from './index';

const meta: Meta<typeof SignInPageContent> = {
  title: 'Components/SignInForm',
  component: SignInPageContent,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;
type Story = StoryObj<typeof SignInPageContent>;

export const Default: Story = {
  render: () => (
    <div className='min-h-screen bg-gray-50 py-12'>
      <SignInPageContent />
    </div>
  ),
};

export const Mobile: Story = {
  parameters: {
    viewport: {
      defaultViewport: 'mobile1',
    },
  },
  render: () => (
    <div className='min-h-screen bg-gray-50 py-12'>
      <SignInPageContent />
    </div>
  ),
};
