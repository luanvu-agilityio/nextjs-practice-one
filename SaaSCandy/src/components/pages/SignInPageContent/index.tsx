'use client';

import { ChangeEvent, Fragment, useState } from 'react';
import Link from 'next/link';
import { usePathname, useRouter } from 'next/navigation';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';

// Components
import { getFriendlyMessage, Heading, showToast } from '@/components/common';
import {
  Breadcrumb,
  BreadcrumbList,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbSeparator,
  BreadcrumbPage,
} from '@/components/common/Breadcrumb';

import { SignInForm, TwoFactorForm } from '@/components/form';

// Icons
import { LogoIcon } from '@/icons';

// Constants
import { AUTH_MESSAGES, ROUTES, TOAST_MESSAGES } from '@/constants';

// Types
import { TOAST_VARIANTS } from '@/types';

// Utils
import { handleSocialAuth } from '@/utils/social-auth';
import { extractBreadcrumbs, SignInFormValues, signInSchema } from '@/utils';

// Service
import { send2FACode, verify2FACode } from '@/service';
import { signIn } from '@/lib/auth-client';

const SignInPageContent = () => {
  const pathname = usePathname();
  const router = useRouter();
  const breadcrumbs = extractBreadcrumbs(pathname);
  const [isLoading, setIsLoading] = useState(false);
  const [requires2FA, setRequires2FA] = useState(false);
  const [twoFactorCode, setTwoFactorCode] = useState('');
  const [userEmail, setUserEmail] = useState('');
  const [userPassword, setUserPassword] = useState('');

  const {
    control,
    handleSubmit,
    formState: { isSubmitting },
  } = useForm<SignInFormValues>({
    resolver: zodResolver(signInSchema),
    mode: 'onBlur',
    reValidateMode: 'onChange',
    defaultValues: { email: '', password: '' },
  });

  const onSubmit = async (data: SignInFormValues) => {
    setIsLoading(true);
    setUserEmail(data.email);
    setUserPassword(data.password);

    try {
      const result = await send2FACode(data.email, data.password);

      if (!result.success) {
        throw new Error(result.error || 'Failed to send verification code');
      }

      setRequires2FA(true);
      setIsLoading(false);

      showToast({
        title: 'Code Sent!',
        description:
          'Please check your email for the 6-digit verification code',
        variant: TOAST_VARIANTS.SUCCESS,
        duration: 5000,
      });

      return;
    } catch (error) {
      console.error('Sign in error:', error);
      showToast({
        title: 'Error',
        description:
          error instanceof Error
            ? error.message
            : 'Failed to send verification code',
        variant: TOAST_VARIANTS.ERROR,
        duration: 5000,
      });
      setIsLoading(false);
    }
  };

  const handleTwoFactorVerification = async () => {
    if (!twoFactorCode || twoFactorCode.length !== 6) {
      showToast({
        title: 'Invalid Code',
        description: 'Please enter a 6-digit code',
        variant: TOAST_VARIANTS.ERROR,
      });
      return;
    }

    setIsLoading(true);
    try {
      // Verify the email code
      const result = await verify2FACode(userEmail, twoFactorCode);

      if (!result.success || !result.data) {
        throw new Error(result.error || 'Invalid verification code');
      }

      const signInResult = await signIn.email({
        email: userEmail,
        password: userPassword,
      });

      if (signInResult.error) {
        throw new Error(signInResult.error.message || 'Sign in failed');
      }

      showToast({
        title: 'Success!',
        description: 'Successfully signed in!',
        variant: TOAST_VARIANTS.SUCCESS,
      });

      setUserPassword('');
      setTwoFactorCode('');
      router.push(ROUTES.HOME);
      router.refresh();
    } catch (error) {
      console.error('2FA verification error:', error);
      showToast({
        title: 'Verification Failed',
        description:
          error instanceof Error ? error.message : 'Invalid or expired code',
        variant: TOAST_VARIANTS.ERROR,
      });
      setIsLoading(false);
    }
  };

  // Resend code functionality
  const handleResendCode = async () => {
    if (!userEmail) return;

    setIsLoading(true);
    try {
      const result = await send2FACode(userEmail, userPassword);

      if (result.success) {
        showToast({
          title: 'Code Resent',
          description: 'A new code has been sent to your email',
          variant: TOAST_VARIANTS.SUCCESS,
        });
        // Reset code input
        setTwoFactorCode('');
      }
    } catch (error) {
      showToast({
        title: 'Error',
        description: getFriendlyMessage(error) ?? 'Failed to resend code',
        variant: TOAST_VARIANTS.ERROR,
      });
    } finally {
      setIsLoading(false);
    }
  };

  const handleSocialSignIn = async (provider: string) => {
    setIsLoading(true);
    try {
      await handleSocialAuth(provider, 'signin');
      showToast({
        title: TOAST_MESSAGES.SOCIAL.SIGNIN_REDIRECT.title,
        description: TOAST_MESSAGES.SOCIAL.SIGNIN_REDIRECT.description.replace(
          '{provider}',
          provider
        ),
        variant: TOAST_VARIANTS.INFO,
        duration: 3000,
      });
    } catch (error) {
      console.error('Social sign-in error:', error);
      showToast({
        title: TOAST_MESSAGES.SOCIAL.SIGNIN_ERROR.title,
        description: TOAST_MESSAGES.SOCIAL.SIGNIN_ERROR.description.replace(
          '{provider}',
          provider
        ),
        variant: TOAST_VARIANTS.ERROR,
        duration: 5000,
      });
      setIsLoading(false);
    }
  };

  const handleForgotPassword = () => {
    router.push(ROUTES.FORGOT_PASSWORD);
  };

  const handleTwoFactorCode = (e: ChangeEvent<HTMLInputElement>) => {
    //  Only allow numbers and limit to 6 digits
    const value = e.target.value.replace(/\D/g, '').slice(0, 6);
    setTwoFactorCode(value);
  };

  const handleBackToSignIn = () => {
    setRequires2FA(false);
    setTwoFactorCode('');
    setUserEmail('');
    setUserPassword('');
  };

  const loading = isSubmitting || isLoading;

  return (
    <div className='flex flex-col items-center max-w-[1296px] mx-auto px-4 sm:px-6'>
      {/* Header Section */}
      <div className='pb-8 sm:pb-12 text-center'>
        <Heading
          as='h2'
          size='xl'
          content={AUTH_MESSAGES.SIGN_IN.title}
          className='text-3xl sm:text-5xl lg:text-6xl'
        />

        <Breadcrumb className='justify-center mt-3 sm:mt-4'>
          <BreadcrumbList>
            {breadcrumbs.map((item, index) => (
              <Fragment key={item.label}>
                <BreadcrumbItem>
                  {item.isActive ? (
                    <BreadcrumbPage className='text-sm sm:xs'>
                      {item.label}
                    </BreadcrumbPage>
                  ) : (
                    <BreadcrumbLink href={item.href} className='text-sm sm:xs'>
                      {item.label}
                    </BreadcrumbLink>
                  )}
                </BreadcrumbItem>
                {index < breadcrumbs.length - 1 && <BreadcrumbSeparator />}
              </Fragment>
            ))}
          </BreadcrumbList>
        </Breadcrumb>
      </div>

      {/* Form Section */}
      <div className='py-8 px-4 sm:py-14 sm:px-16 w-full lg:w-159 rounded-2xl sm:rounded-4xl border border-form-border-color shadow-form'>
        <div className='flex flex-col gap-6 sm:gap-10'>
          {/* Logo */}
          <div className='flex justify-center'>
            <Link
              href={ROUTES.HOME}
              className='flex items-center gap-2 sm:gap-3'
            >
              <LogoIcon className='w-8 h-8 sm:w-10 sm:h-10' />
              <span className='text-2xl sm:text-3xl font-secondary text-primary'>
                SaaS<span className='font-medium'>Candy</span>
              </span>
            </Link>
          </div>

          <div className='flex flex-col gap-6 sm:gap-8'>
            {/* Toggle between SignIn and 2FA forms */}
            {!requires2FA ? (
              <SignInForm
                control={control}
                onSubmit={handleSubmit(onSubmit)}
                onSocialSignIn={handleSocialSignIn}
                isLoading={loading}
              />
            ) : (
              <TwoFactorForm
                userEmail={userEmail}
                twoFactorCode={twoFactorCode}
                onCodeChange={handleTwoFactorCode}
                onVerify={handleTwoFactorVerification}
                onResendCode={handleResendCode}
                onBack={handleBackToSignIn}
                isLoading={loading}
              />
            )}

            {/* Footer Links */}
            <div className='text-center space-y-2 sm:space-y-3'>
              {!requires2FA && (
                <button
                  onClick={handleForgotPassword}
                  className='text-sm sm:text-md font-regular text-primary hover:text-primary/80 transition-colors'
                  disabled={loading}
                >
                  {AUTH_MESSAGES.SIGN_IN.forgotPassword}
                </button>
              )}

              <div className='text-sm sm:text-md font-regular text-primary'>
                {AUTH_MESSAGES.SIGN_IN.notMember}{' '}
                <Link
                  href={ROUTES.SIGN_UP}
                  className='text-primary font-medium hover:underline'
                >
                  {AUTH_MESSAGES.SIGN_IN.signUpLink}
                </Link>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export { SignInPageContent };
