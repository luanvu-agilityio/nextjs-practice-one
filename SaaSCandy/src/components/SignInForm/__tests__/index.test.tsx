import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SignInPageContent from '../index';

const mockPush = jest.fn();
const mockPathname = '/signin';
const mockSignIn = jest.fn();
const mockSignInAction = jest.fn();
const mockHandleSocialAuth = jest.fn();
const mockShowToast = jest.fn();

// Mock useActionState
const mockUseActionState = jest.fn();

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useActionState: (...args: unknown[]) => mockUseActionState(...args),
}));

jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
  usePathname: () => mockPathname,
}));

jest.mock('next/link', () => {
  const MockedLink = ({
    children,
    href,
  }: {
    children: React.ReactNode;
    href: string;
  }) => {
    return <a href={href}>{children}</a>;
  };
  MockedLink.displayName = 'MockedLink';
  return MockedLink;
});

jest.mock('better-auth/react', () => ({
  signIn: (...args: unknown[]) => mockSignIn(...args),
}));

jest.mock('@/lib/auth', () => ({
  signInAction: (...args: unknown[]) => mockSignInAction(...args),
}));

jest.mock('@/utils/auth', () => ({
  extractBreadcrumbs: jest.fn(() => [
    { label: 'Home', href: '/', isActive: false },
    { label: 'Sign In', href: '/signin', isActive: true },
  ]),
  handleSocialAuth: (...args: unknown[]) => mockHandleSocialAuth(...args),
}));

// Mock Toast component
jest.mock('@/components/common/Toast', () => ({
  showToast: (...args: unknown[]) => mockShowToast(...args),
}));

jest.mock('@/constants', () => ({
  AUTH_MESSAGES: {
    SIGN_IN: {
      title: 'Sign In',
      submitButton: 'Sign In',
      submittingButton: 'Signing In...',
      forgotPassword: 'Forgot Password?',
      notMember: 'Not a member?',
      signUpLink: 'Sign up',
    },
    SOCIAL: {
      googleSignIn: 'Sign in with Google',
      githubSignIn: 'Sign in with GitHub',
    },
    DIVIDER: 'or',
  },
  ROUTES: {
    HOME: '/',
    SIGN_IN: '/signin',
    SIGN_UP: '/signup',
  },
  SOCIAL_PROVIDERS: {
    GOOGLE: 'google',
    GITHUB: 'github',
  },
  TOAST_MESSAGES: {
    SIGN_IN: {
      SUCCESS: {
        title: 'Sign In Successful!',
        description: 'Welcome back! Redirecting to your dashboard...',
      },
      ERROR: {
        title: 'Sign In Failed',
        description: 'Invalid email or password. Please try again.',
      },
      AUTH_ERROR: {
        title: 'Authentication Error',
        description:
          'Sign in successful but session creation failed. Please try again.',
      },
    },
    SOCIAL: {
      SIGNIN_REDIRECT: {
        title: 'Redirecting...',
        description: 'Signing in with {provider}. Please wait...',
      },
      SIGNIN_ERROR: {
        title: 'Social Sign In Failed',
        description: 'Failed to sign in with {provider}. Please try again.',
      },
    },
    FEATURES: {
      FORGOT_PASSWORD: {
        title: 'Feature Coming Soon',
        description: 'Password reset functionality will be available soon.',
      },
    },
  },
}));

// Mock React Hook Form
jest.mock('react-hook-form', () => ({
  useForm: jest.fn(() => ({
    control: {},
    handleSubmit: jest.fn(fn => (e: Event) => {
      e.preventDefault();
      fn({ email: 'test@example.com', password: 'password123' });
    }),
    formState: {
      isSubmitting: false,
      errors: {},
    },
    clearErrors: jest.fn(),
  })),
  Controller: ({
    render,
  }: {
    render: (props: unknown) => React.ReactElement;
  }) =>
    render({
      field: { onChange: jest.fn(), value: '', name: 'test' },
      fieldState: { error: null },
    }),
}));

jest.mock('@hookform/resolvers/zod', () => ({
  zodResolver: jest.fn(),
}));

jest.mock('@/components/common', () => ({
  InputController: ({
    name,
    placeholder,
    type,
  }: {
    name: string;
    placeholder?: string;
    type?: string;
  }) => (
    <input
      data-testid={name}
      name={name}
      placeholder={placeholder}
      type={type}
    />
  ),
  Button: ({
    children,
    disabled,
    type,
    ...props
  }: {
    children: React.ReactNode;
    type?: 'button' | 'reset' | 'submit';
    disabled?: boolean;
    [key: string]: unknown;
  }) => (
    <button type={type} disabled={disabled} {...props}>
      {children}
    </button>
  ),
  Heading: ({ content }: { content: string }) => <h2>{content}</h2>,
  Divider: ({ text }: { text: string }) => <div>{text}</div>,
}));

jest.mock('@/components/common/Breadcrumb', () => ({
  Breadcrumb: ({ children }: { children: React.ReactNode }) => (
    <nav>{children}</nav>
  ),
  BreadcrumbList: ({ children }: { children: React.ReactNode }) => (
    <ol>{children}</ol>
  ),
  BreadcrumbItem: ({ children }: { children: React.ReactNode }) => (
    <li>{children}</li>
  ),
  BreadcrumbLink: ({
    children,
    href,
  }: {
    children: React.ReactNode;
    href: string;
  }) => <a href={href}>{children}</a>,
  BreadcrumbSeparator: () => <span>/</span>,
  BreadcrumbPage: ({ children }: { children: React.ReactNode }) => (
    <span>{children}</span>
  ),
}));

jest.mock('@/components/SocialButton', () => {
  return function SocialButton({
    children,
    provider,
    onClick,
  }: {
    children: React.ReactNode;
    provider: string;
    onClick: () => void;
  }) {
    return (
      <button onClick={onClick} data-provider={provider}>
        {children}
      </button>
    );
  };
});

jest.mock('@/components/icons/GitHubIcon', () => {
  return function GitHubIcon() {
    return <svg data-testid='github-icon' />;
  };
});

jest.mock('@/components/icons/GoogleIcon', () => {
  return function GoogleIcon() {
    return <svg data-testid='google-icon' />;
  };
});

jest.mock('@/components/icons/Logo', () => {
  return function LogoIcon({ className }: { className?: string }) {
    return <svg data-testid='logo-icon' className={className} />;
  };
});

jest.mock('@/utils', () => ({
  signInSchema: jest.requireActual('zod').z.object({
    email: jest.requireActual('zod').z.string().email(),
    password: jest.requireActual('zod').z.string().min(1),
  }),
}));

describe('SignInForm - Snapshot', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    mockUseActionState.mockReturnValue([{ error: null }, jest.fn(), false]);
  });

  it('matches snapshot', () => {
    const { container } = render(<SignInPageContent />);
    expect(container).toMatchSnapshot();
  });
});

describe('SignInForm - Interactive', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    mockUseActionState.mockReturnValue([{ error: null }, jest.fn(), false]);
  });

  it('handles social sign in button clicks', async () => {
    const user = userEvent.setup();

    render(<SignInPageContent />);

    const googleButton = screen.getByRole('button', {
      name: /sign in with google/i,
    });
    await user.click(googleButton);

    await waitFor(() => {
      expect(mockHandleSocialAuth).toHaveBeenCalledWith('google', 'signin');
    });
  });
});
