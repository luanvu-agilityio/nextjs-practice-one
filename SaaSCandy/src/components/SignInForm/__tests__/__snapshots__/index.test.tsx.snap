// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SignInForm - Snapshot matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center max-w-[1296px] mx-auto px-4 sm:px-6"
  >
    <div
      class="pb-8 sm:pb-12 text-center"
    >
      <h2>
        Sign In
      </h2>
      <nav>
        <ol>
          <li>
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <span>
            /
          </span>
          <li>
            <span>
              Sign In
            </span>
          </li>
        </ol>
      </nav>
    </div>
    <div
      class="py-8 px-4 sm:py-14 sm:px-16 w-full lg:w-159 rounded-2xl sm:rounded-4xl border border-form-border-color shadow-form"
    >
      <div
        class="flex flex-col gap-6 sm:gap-10"
      >
        <div
          class="flex justify-center"
        >
          <a
            href="/"
          >
            <svg
              class="w-8 h-8 sm:w-10 sm:h-10"
              data-testid="logo-icon"
            />
            <span
              class="text-2xl sm:text-3xl font-secondary text-primary"
            >
              SaaS
              <span
                class="font-medium"
              >
                Candy
              </span>
            </span>
          </a>
        </div>
        <div
          class="flex flex-col gap-6 sm:gap-8"
        >
          <div
            class="flex flex-col sm:flex-row gap-3 sm:gap-4"
          >
            <button
              data-provider="google"
            >
              <span
                class="hidden sm:inline"
              >
                Sign in with Google
              </span>
              <span
                class="sm:hidden"
              >
                Google
              </span>
            </button>
            <button
              data-provider="github"
            >
              <span
                class="hidden sm:inline"
              >
                Sign in with GitHub
              </span>
              <span
                class="sm:hidden"
              >
                GitHub
              </span>
            </button>
          </div>
          <div>
            or
          </div>
          <form
            class="space-y-3 sm:space-y-4"
            novalidate=""
          >
            <input
              data-testid="email"
              name="email"
              placeholder="Email"
              type="email"
            />
            <input
              data-testid="password"
              name="password"
              placeholder="Password"
              type="password"
            />
            <button
              class="w-full"
              type="submit"
              variant="primary"
            >
              Sign In
            </button>
          </form>
          <div
            class="text-center space-y-2 sm:space-y-3"
          >
            <button
              class="text-sm sm:text-md font-regular text-primary hover:text-primary/80 transition-colors"
            >
              Forgot Password?
            </button>
            <div
              class="text-sm sm:text-md font-regular text-primary"
            >
              Not a member?
               
              <a
                href="/signup"
              >
                Sign up
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
